File: MyMusic.Mobile/src/services/sync/phases.ts
import type { SyncDeps, SyncContext, SyncFileInfo, ScanError, SyncConflict, ProgressHandler, SyncRecordItem, ActionResult } from './types';
import { SyncActionCounts, addDeltaToResult } from './types';
import type { SyncPotentialConflictItem, SyncPotentialUpdateItem, RenameData } from '../../api/types';
import { SyncCancelledError } from './errors';
import { safeToIsoString, chunkArray, formatFilePath } from './utils';
import { actionCreateRemote, actionUpdateRemote, actionCreateLocal, actionDelete, actionConflict, actionRename } from './sync-actions-device';

const EMPTY_COUNTS: SyncActionCounts = {
    createRemoteCount: 0,
    updateRemoteCount: 0,
    skippedCount: 0,
    createLocalCount: 0,
    updateLocalCount: 0,
    deleteCount: 0,
    linkCount: 0,
    unlinkCount: 0,
    renameCount: 0,
    conflictCount: 0,
    updateTimestampCount: 0,
    errorCount: 0,
};

export interface ScanPhaseResult {
    files: SyncFileInfo[];
    errors: ScanError[];
    estimatedTotal: number;
}

export async function scanPhase (
    deps: SyncDeps,
    ctx: SyncContext,
    onProgress: ProgressHandler,
    previousScanTotal: number | null
): Promise<ScanPhaseResult> {
    const estimatedTotal = previousScanTotal || 0;

    onProgress({
        phase: 'scanning',
        totalFiles: 0,
        estimatedTotalFiles: estimatedTotal,
        processedFiles: 0,
        scannedFiles: 0,
        currentFile: 'Scanning your music folder...',
    });

    const scanErrors: ScanError[] = [];
    let currentEstimate = estimatedTotal;

    const { files, errors } = await deps.scanner(ctx.repositoryPath, {
        extensions: deps.config.getMusicExtensions(),
        excludePatterns: deps.config.getExcludePatterns(),
        basePath: ctx.repositoryPath,
        onProgress: (scannedCount, currentDir) => {
            if (deps.state.isCancelled) {
                throw new SyncCancelledError();
            }
            const dirName = currentDir.split('/').pop() || 'music folder';
            if (scannedCount > currentEstimate) {
                currentEstimate = scannedCount;
            }
            onProgress({
                scannedFiles: scannedCount,
                estimatedTotalFiles: currentEstimate,
                currentFile: `${scannedCount} files found in ${dirName}...`,
            });
        },
        onError: (path, error) => {
            scanErrors.push({ path, error });
        },
    });

    ctx.result.error += errors.length;

    if (deps.state.isCancelled) {
        throw new SyncCancelledError();
    }

    if (files.length > currentEstimate) {
        currentEstimate = files.length;
    }

    onProgress({
        totalFiles: files.length,
        estimatedTotalFiles: currentEstimate,
        scannedFiles: files.length,
        phase: 'upload',
        currentFile: '',
    });

    return { files, errors: scanErrors, estimatedTotal: currentEstimate };
}

export async function startSessionPhase (
    deps: SyncDeps,
    ctx: SyncContext,
    scanErrors: ScanError[],
    onProgress: ProgressHandler
): Promise<void> {
    const startResponse = await deps.apiClient.startSync(ctx.deviceId, {
        dryRun: ctx.options.dryRun,
        repositoryPath: ctx.repositoryPath,
        scanErrors: scanErrors.map(e => ({ path: e.path, error: e.error })),
    });
    ctx.sessionId = startResponse.sessionId;

    onProgress({ phase: 'server' });
}

export async function resolveConflictsPhase (
    deps: SyncDeps,
    ctx: SyncContext,
    potentialConflicts: SyncPotentialConflictItem[],
    potentialUpdates: SyncPotentialUpdateItem[],
    chunk: SyncFileInfo[],
    toUpdatePaths: Set<string>,
    onProgress: ProgressHandler
): Promise<void> {
    await actionConflict(
        deps.apiClient,
        deps.fileOps,
        deps.userPrompt,
        ctx,
        potentialConflicts,
        potentialUpdates,
        chunk,
        toUpdatePaths,
        (progress) => {
            onProgress(progress);
        }
    );

    for (const conflict of potentialConflicts) {
        if (conflict.songId != null && !toUpdatePaths.has(conflict.path)) {
            ctx.conflictedSongIds.add(conflict.songId);
        }
    }
}

export async function uploadPhase (
    deps: SyncDeps,
    ctx: SyncContext,
    files: SyncFileInfo[],
    onProgress: ProgressHandler
): Promise<void> {
    if (files.length === 0) {
        return;
    }

    const chunkSize = deps.config.getChunkSize();
    const chunks = chunkArray(files, chunkSize);

    for (let i = 0; i < chunks.length; i++) {
        if (deps.state.isCancelled) {
            throw new SyncCancelledError();
        }
        const chunk = chunks[i];

        const syncFiles = chunk.map(f => ({
            path: f.relativePath,
            modifiedAt: safeToIsoString(f.modifiedAt)!,
            createdAt: safeToIsoString(f.createdAt)!,
        }));

        let syncResponse;
        try {
            syncResponse = await deps.apiClient.checkSync(ctx.deviceId, ctx.sessionId!, {
                files: syncFiles,
                force: ctx.options.force,
            });
        } catch (e) {
            ctx.result.error += chunk.length;
            onProgress({
                phase: 'upload',
                errorMessage: `Chunk ${i + 1} failed: ${e instanceof Error ? e.message : String(e)}`,
            });
            continue;
        }

        ctx.result = addDeltaToResult(ctx.result, syncResponse.counts ?? EMPTY_COUNTS);

        if (syncResponse.records.length > 0) {
            ctx.pendingActions = mergePendingActions(ctx.pendingActions ?? [], syncResponse.records);
            ctx.pendingDownloadPaths = extractPendingDownloadPaths(ctx.pendingActions);
        }

        const toCreatePaths = new Set(syncResponse.toCreate.map(f => f.path));
        const toUpdatePaths = new Set(syncResponse.toUpdate.map(f => f.path));

        if (syncResponse.potentialConflicts.length > 0 || syncResponse.potentialUpdates.length > 0) {
            await actionConflict(
                deps.apiClient,
                deps.fileOps,
                deps.userPrompt,
                ctx,
                syncResponse.potentialConflicts,
                syncResponse.potentialUpdates,
                chunk,
                toUpdatePaths,
                (progress) => {
                    onProgress({
                        phase: progress.phase ?? 'resolving',
                        currentFile: progress.currentFile,
                        conflict: progress.conflict,
                    });
                }
            );

            for (const conflict of syncResponse.potentialConflicts) {
                if (conflict.songId != null && !toUpdatePaths.has(conflict.path)) {
                    ctx.conflictedSongIds.add(conflict.songId);
                }
            }
        }

        await processChunkUploads(deps, ctx, chunk, syncResponse.toCreate, syncResponse.toUpdate, toUpdatePaths, onProgress);
    }
}

export async function serverActionsPhase (
    deps: SyncDeps,
    ctx: SyncContext,
    onProgress: ProgressHandler
): Promise<ActionResult[]> {
    if (deps.state.isCancelled) {
        throw new SyncCancelledError();
    }

    const pendingActionsResponse = await deps.apiClient.createPendingActions(ctx.deviceId, ctx.sessionId!);
    ctx.pendingActions = mergePendingActions(ctx.pendingActions ?? [], pendingActionsResponse.records);
    ctx.pendingDownloadPaths = extractPendingDownloadPaths(ctx.pendingActions);

    const pendingActions = ctx.pendingActions;
    const serverResults: ActionResult[] = [];

    for (const record of pendingActions) {
        if (deps.state.isCancelled) {
            throw new SyncCancelledError();
        }

        if (ctx.uploadedPaths.has(record.filePath) && record.action !== 'CreateLocal' && record.action !== 'UpdateLocal') {
            await deps.apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, {
                recordIds: [record.id],
            });
            continue;
        }

        if ((record.action === 'CreateLocal' || record.action === 'UpdateLocal')) {
            if (record.songId != null && ctx.conflictedSongIds.has(record.songId)) {
                continue;
            }

            const result = await actionCreateLocal(
                deps.apiClient,
                deps.fileOps,
                deps.userPrompt,
                ctx,
                record.songId,
                record.filePath,
                ctx.decodedRepoPath,
                record.id,
                record.reason ?? undefined
            );
            if (result) {
                serverResults.push(result);
                if (result.counts) {
                    ctx.result = addDeltaToResult(ctx.result, result.counts);
                }
            }
        } else if (record.action === 'Unlink' || record.action === 'Delete') {
            const result = await actionDelete(
                deps.apiClient,
                deps.fileOps,
                deps.userPrompt,
                ctx,
                record.filePath,
                ctx.decodedRepoPath,
                record.songId ?? undefined,
                record.id,
                record.reason ?? undefined
            );
            if (result) {
                serverResults.push(result);
                if (result.counts) {
                    ctx.result = addDeltaToResult(ctx.result, result.counts);
                }
            }
        } else if (record.action === 'Rename') {
            const renameData = record.data;
            if (renameData?.previousPath) {
                const result = await actionRename(
                    deps.apiClient,
                    deps.fileOps,
                    ctx,
                    record.filePath,
                    renameData.previousPath,
                    ctx.decodedRepoPath,
                    record.id
                );
                if (result) {
                    serverResults.push(result);
                    if (result.counts) {
                        ctx.result = addDeltaToResult(ctx.result, result.counts);
                    }
                }
            }
        }

        onProgress({
            createLocal: ctx.result.createLocal,
            delete: ctx.result.delete,
            processedFiles: pendingActions.indexOf(record) + 1,
            totalFiles: pendingActions.length,
        });
    }

    return serverResults;
}

export async function commitPhase (
    deps: SyncDeps,
    ctx: SyncContext,
    onProgress: ProgressHandler
): Promise<void> {
    if (deps.state.isCancelled) {
        throw new SyncCancelledError();
    }

    onProgress({ phase: 'committing' });

    const commitResponse = await deps.apiClient.commitSync(ctx.deviceId, ctx.sessionId!);

    ctx.result.createRemote = commitResponse.createRemoteCount;
    ctx.result.updateRemote = commitResponse.updateRemoteCount;
    ctx.result.skipped = commitResponse.skippedCount;
    ctx.result.createLocal = commitResponse.createLocalCount;
    ctx.result.updateLocal = commitResponse.updateLocalCount;
    ctx.result.delete = commitResponse.deleteCount;
    ctx.result.link = commitResponse.linkCount;
    ctx.result.unlink = commitResponse.unlinkCount;
    ctx.result.rename = commitResponse.renameCount;
    ctx.result.conflict = commitResponse.conflictCount;
    ctx.result.updateTimestamp = commitResponse.updateTimestampCount;
    ctx.result.error = commitResponse.errorCount;
}

export async function completePhase (
    deps: SyncDeps,
    ctx: SyncContext,
    filesCount: number,
    onProgress: ProgressHandler
): Promise<void> {
    onProgress({ phase: 'completing' });

    const completeResponse = await deps.apiClient.completeSync(ctx.deviceId, ctx.sessionId!);

    ctx.result.createRemote = completeResponse.createRemoteCount;
    ctx.result.updateRemote = completeResponse.updateRemoteCount;
    ctx.result.skipped = completeResponse.skippedCount;
    ctx.result.createLocal = completeResponse.createLocalCount;
    ctx.result.updateLocal = completeResponse.updateLocalCount;
    ctx.result.delete = completeResponse.deleteCount;
    ctx.result.link = completeResponse.linkCount;
    ctx.result.unlink = completeResponse.unlinkCount;
    ctx.result.rename = completeResponse.renameCount;
    ctx.result.conflict = completeResponse.conflictCount;
    ctx.result.updateTimestamp = completeResponse.updateTimestampCount;
    ctx.result.error = completeResponse.errorCount;
    ctx.result.sessionId = ctx.sessionId;

    await deps.config.setLastSyncAt(new Date().toISOString());
    await deps.config.setLastScanTotal(filesCount);
}

async function processChunkUploads (
    deps: SyncDeps,
    ctx: SyncContext,
    chunk: SyncFileInfo[],
    toCreate: Array<{ path: string; modifiedAt: Date; createdAt: Date; reason?: string; }>,
    toUpdate: Array<{ path: string; modifiedAt: Date; createdAt: Date; reason?: string; }>,
    toUpdatePaths: Set<string>,
    onProgress: ProgressHandler
): Promise<void> {
    for (const fileToCreate of toCreate) {
        if (deps.state.isCancelled) {
            throw new SyncCancelledError();
        }

        const file = chunk.find(f => f.relativePath === fileToCreate.path);
        if (file) {
            const result = await actionCreateRemote(deps.apiClient, deps.fileOps, ctx, file, fileToCreate.reason);
            if (result.counts) {
                ctx.result = addDeltaToResult(ctx.result, result.counts);
            }
        }
        ctx.uploadedPaths.add(fileToCreate.path);

        onProgress({
            processedFiles: ctx.result.createRemote + ctx.result.updateRemote + ctx.result.skipped + ctx.result.error + ctx.result.conflict,
            currentFile: formatFilePath(fileToCreate.path, ctx.repositoryPath),
            createRemote: ctx.result.createRemote,
            updateRemote: ctx.result.updateRemote,
            skipped: ctx.result.skipped,
            error: ctx.result.error,
            conflict: ctx.result.conflict,
        });
    }

    for (const fileToUpdate of toUpdate) {
        if (deps.state.isCancelled) {
            throw new SyncCancelledError();
        }

        if (!toUpdatePaths.has(fileToUpdate.path)) {
            continue;
        }

        const file = chunk.find(f => f.relativePath === fileToUpdate.path);
        if (file) {
            const result = await actionUpdateRemote(deps.apiClient, deps.fileOps, ctx, file, fileToUpdate.reason);
            if (result.counts) {
                ctx.result = addDeltaToResult(ctx.result, result.counts);
            }
        }
        ctx.uploadedPaths.add(fileToUpdate.path);

        onProgress({
            processedFiles: ctx.result.createRemote + ctx.result.updateRemote + ctx.result.skipped + ctx.result.error + ctx.result.conflict,
            currentFile: formatFilePath(fileToUpdate.path, ctx.repositoryPath),
            createRemote: ctx.result.createRemote,
            updateRemote: ctx.result.updateRemote,
            skipped: ctx.result.skipped,
            error: ctx.result.error,
            conflict: ctx.result.conflict,
        });
    }
}

function extractPendingDownloadPaths (records: SyncRecordItem[]): Set<string> {
    const paths = new Set<string>();
    for (const record of records) {
        if (record.action === 'CreateLocal' || record.action === 'UpdateLocal') {
            paths.add(record.filePath);
        }
    }
    return paths;
}

function mergePendingActions (
    existing: SyncRecordItem[],
    incoming: SyncRecordItem[]
): SyncRecordItem[] {
    const merged = new Map<number, SyncRecordItem>();
    for (const record of existing) {
        merged.set(record.id, record);
    }
    for (const record of incoming) {
        merged.set(record.id, record);
    }
    return Array.from(merged.values());
}

File: MyMusic.CLI/Services/Sync/Phases.cs
namespace MyMusic.CLI.Services.Sync;

using System.Text.Json;
using Microsoft.Extensions.Logging;
using MyMusic.CLI.Services.Sync.Types;

public class Phases(
    ISyncApiClient apiClient,
    SyncActionsDevice syncActions,
    ISyncConfig config,
    IFileSystemScanner scanner,
    ILogger<Phases> logger)
{
    public async Task<ScanResult> ScanPhaseAsync(
        SyncContext ctx,
        IProgress<SyncProgress>? progress,
        CancellationToken ct = default)
    {
        logger.LogInformation("Scanning repository: {Path}", ctx.RepositoryPath);

        progress?.Report(SyncProgress.ForPhase("scanning", "Scanning your music folder..."));

        var scanErrors = new List<ScanError>();

        var result = await scanner.ScanAsync(
            ctx.RepositoryPath,
            config.GetMusicExtensions(),
            config.GetExcludePatterns(),
            onProgress: (scannedCount, currentDir) =>
            {
                progress?.Report(SyncProgress.ForScanProgress(scannedCount, currentDir));
            },
            onError: (path, error) =>
            {
                scanErrors.Add(new ScanError { Path = path, Error = error });
            },
            ct);

        logger.LogInformation("Found {Count} music files", result.Files.Count);

        if (scanErrors.Count > 0)
        {
            logger.LogWarning("Scan completed with {Count} errors", scanErrors.Count);
            ctx.Result = ctx.Result.AddDelta(new SyncActionCounts { ErrorCount = scanErrors.Count });
        }

        return result;
    }

    public async Task StartSessionAsync(
        SyncContext ctx,
        List<ScanError> scanErrors,
        CancellationToken ct = default)
    {
        var startResponse = await apiClient.StartSyncAsync(ctx.DeviceId, new StartSyncRequest
        {
            DryRun = ctx.Options.DryRun,
            RepositoryPath = ctx.RepositoryPath,
            ScanErrors = scanErrors
        }, ct);
        ctx.SessionId = startResponse.SessionId;
        logger.LogInformation("Started sync session: {SessionId} (DryRun: {DryRun})", ctx.SessionId, ctx.Options.DryRun);
    }

    public async Task UploadPhaseAsync(
        SyncContext ctx,
        List<ScannedFile> files,
        IProgress<SyncProgress>? progress,
        CancellationToken ct = default)
    {
        if (ctx.Options.Direction == SyncDirection.Down)
        {
            logger.LogInformation("Skipping upload phase (direction: down)");
            return;
        }

        if (files.Count == 0)
        {
            return;
        }

        var chunkSize = config.GetChunkSize();
        var chunks = files.Chunk(chunkSize).ToList();
        logger.LogInformation("Processing {ChunkCount} chunks", chunks.Count);

        for (var i = 0; i < chunks.Count; i++)
        {
            if (ct.IsCancellationRequested)
            {
                break;
            }

            var chunk = chunks[i];
            var chunkNumber = i + 1;
            logger.LogInformation("Processing chunk {ChunkNumber}/{TotalChunks}", chunkNumber, chunks.Count);

            var syncFiles = chunk
                .Select(f => new SyncFileInfo
                {
                    Path = f.RelativePath,
                    ModifiedAt = f.ModifiedAt,
                    CreatedAt = f.CreatedAt
                }).ToList();

            var syncRequest = new CheckSyncRequest
            {
                Files = syncFiles,
                Force = ctx.Options.Force
            };

            CheckSyncResult syncResponse;
            try
            {
                syncResponse = await apiClient.CheckSyncAsync(ctx.DeviceId, ctx.SessionId, syncRequest, ct);
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "Failed to check sync for chunk {ChunkNumber}", chunkNumber);
                ctx.Result = ctx.Result.AddDelta(new SyncActionCounts { ErrorCount = chunk.Length });
                progress?.Report(SyncProgress.FromResult(
                    ctx.Result, "upload", files.Count, (i + 1) * chunkSize,
                    errorMessage: $"Chunk {chunkNumber} failed: {ex.Message}"));
                continue;
            }

            ctx.Result = ctx.Result.AddDelta(syncResponse.Counts);

            if (syncResponse.Records.Count > 0)
            {
                ctx.PendingServerRecords.AddRange(syncResponse.Records);
                logger.LogInformation(
                    "Chunk {ChunkNumber}: {RecordCount} server action records accumulated",
                    chunkNumber, syncResponse.Records.Count);
            }

            var conflictRecords = syncResponse.Records.Where(r => r.Action == SyncRecordAction.Conflict).ToList();
            var updateLocalRecords = syncResponse.Records.Where(r => r.Action == SyncRecordAction.UpdateLocal).ToList();

            if (conflictRecords.Count > 0 || updateLocalRecords.Count > 0)
            {
                await ResolveConflictsAsync(ctx, conflictRecords, updateLocalRecords, ct);

                var superseded = conflictRecords
                    .Concat(updateLocalRecords)
                    .ToHashSet(ReferenceEqualityComparer.Instance);
                ctx.PendingServerRecords.RemoveAll(superseded.Contains);
            }

            var toCreateRecords = syncResponse.Records.Where(r => r.Action == SyncRecordAction.CreateRemote).ToList();
            var toUpdatePaths = new HashSet<string>();
            foreach (var updateRecord in syncResponse.Records.Where(r => r.Action == SyncRecordAction.UpdateRemote))
            {
                toUpdatePaths.Add(updateRecord.FilePath);
            }

            logger.LogInformation(
                "Chunk {ChunkNumber}: {ToCreate} to create, {ToUpdate} to update, {Conflicts} conflicts, {PotentialUpdates} potential updates",
                chunkNumber, toCreateRecords.Count, toUpdatePaths.Count,
                conflictRecords.Count, updateLocalRecords.Count);

            foreach (var createRecord in toCreateRecords)
            {
                if (ct.IsCancellationRequested) break;

                var createData = createRecord.Data.HasValue ? DeserializeCheckCreateUpdateData(createRecord.Data) : null;
                var fileInfo = new SyncFileInfo
                {
                    Path = createRecord.FilePath,
                    ModifiedAt = createData?.ModifiedAt ?? DateTime.MinValue,
                    CreatedAt = createData?.CreatedAt ?? DateTime.MinValue,
                    Reason = createRecord.Reason
                };

                var result = await syncActions.ActionCreateRemoteAsync(
                    ctx.DeviceId, ctx.SessionId, ctx.RepositoryPath, fileInfo, ct);

                if (result.Counts != null)
                {
                    ctx.Result = ctx.Result.AddDelta(result.Counts);
                }

                ctx.UploadedPaths.Add(createRecord.FilePath);
                progress?.Report(SyncProgress.FromResult(
                    ctx.Result, "upload", files.Count, 0,
                    createRecord.FilePath, result.Action == "Error" ? result.ErrorMessage : null));
            }

            foreach (var updateRecord in syncResponse.Records.Where(r => r.Action == SyncRecordAction.UpdateRemote))
            {
                if (ct.IsCancellationRequested) break;

                if (!toUpdatePaths.Contains(updateRecord.FilePath))
                {
                    continue;
                }

                var updateData = updateRecord.Data.HasValue ? DeserializeCheckCreateUpdateData(updateRecord.Data) : null;
                var fileInfo = new SyncFileInfo
                {
                    Path = updateRecord.FilePath,
                    ModifiedAt = updateData?.ModifiedAt ?? DateTime.MinValue,
                    CreatedAt = updateData?.CreatedAt ?? DateTime.MinValue,
                    Reason = updateRecord.Reason
                };

                var result = await syncActions.ActionUpdateRemoteAsync(
                    ctx.DeviceId, ctx.SessionId, ctx.RepositoryPath, fileInfo, ct);

                if (result.Counts != null)
                {
                    ctx.Result = ctx.Result.AddDelta(result.Counts);
                }

                ctx.UploadedPaths.Add(updateRecord.FilePath);
                progress?.Report(SyncProgress.FromResult(
                    ctx.Result, "upload", files.Count, 0,
                    updateRecord.FilePath, result.Action == "Error" ? result.ErrorMessage : null));
            }
        }
    }

    public async Task ResolveConflictsAsync(
        SyncContext ctx,
        List<SyncRecordItem> conflictRecords,
        List<SyncRecordItem> updateLocalRecords,
        CancellationToken ct = default)
    {
        if (conflictRecords.Count > 0 || updateLocalRecords.Count > 0)
        {
            var result = await syncActions.ActionConflictAsync(
                ctx.DeviceId, ctx.SessionId, ctx.RepositoryPath, conflictRecords, updateLocalRecords, ct);

            ctx.Result = ctx.Result.AddDelta(result.Counts);

            var unresolvedSongIds = result.Records
                .Where(r => (r.Action == SyncRecordAction.Conflict || r.Action == SyncRecordAction.Error) && r.SongId.HasValue)
                .Select(r => r.SongId!.Value)
                .ToHashSet();

            ctx.ConflictedSongIds.Clear();
            foreach (var songId in unresolvedSongIds)
            {
                ctx.ConflictedSongIds.Add(songId);
            }

            foreach (var record in result.Records)
            {
                if (record.Action == SyncRecordAction.UpdateLocal ||
                    record.Action == SyncRecordAction.Rename)
                {
                    ctx.PendingServerRecords.Add(record);
                }
            }
        }
    }

    public async Task ServerActionsPhaseAsync(
        SyncContext ctx,
        IProgress<SyncProgress>? progress,
        CancellationToken ct = default)
    {
        if (ctx.Options.Direction == SyncDirection.Up)
        {
            logger.LogInformation("Skipping server actions phase (direction: up)");
            return;
        }

        List<SyncRecordItem> recordsToProcess;

        logger.LogInformation("Fetching pending actions for device {DeviceId}", ctx.DeviceId);
        var pendingResponse = await apiClient.CreatePendingActionsAsync(ctx.DeviceId, ctx.SessionId, ct);

        var existingIds = ctx.PendingServerRecords.Select(r => r.Id).ToHashSet();
        var newRecords = pendingResponse.Records
            .Where(r => !existingIds.Contains(r.Id))
            .ToList();
        ctx.PendingServerRecords.AddRange(newRecords);

        recordsToProcess = ctx.PendingServerRecords;

        if (ctx.Options.Direction == SyncDirection.Down)
        {
            recordsToProcess = recordsToProcess
                .Where(r => r.Action == SyncRecordAction.CreateLocal || r.Action == SyncRecordAction.UpdateLocal)
                .ToList();
            logger.LogInformation("Found {Count} songs to download", recordsToProcess.Count);
        }
        else
        {
            logger.LogInformation("Processing {Count} pending actions", recordsToProcess.Count);
        }

        var serverTotal = recordsToProcess.Count;

        foreach (var record in recordsToProcess)
        {
            if (ct.IsCancellationRequested) break;

            var fullPath = Path.Combine(ctx.RepositoryPath, record.FilePath);

            if (ctx.UploadedPaths.Contains(record.FilePath) && record.Action != SyncRecordAction.CreateLocal && record.Action != SyncRecordAction.UpdateLocal)
            {
                await apiClient.AcknowledgeActionAsync(ctx.DeviceId, ctx.SessionId, new AcknowledgeActionRequest
                {
                    RecordIds = [record.Id]
                }, ct);
                continue;
            }

            if ((record.Action == SyncRecordAction.CreateLocal || record.Action == SyncRecordAction.UpdateLocal) && record.SongId.HasValue && ctx.ConflictedSongIds.Contains(record.SongId.Value))
            {
                logger.LogInformation("Skipping download for song {SongId} - unresolved conflict", record.SongId);
                continue;
            }

            if (record.Action == SyncRecordAction.CreateLocal)
            {
                logger.LogInformation("Creating local song {SongId} at {Path}", record.SongId, record.FilePath);

                var result = await syncActions.ActionCreateLocalAsync(
                    ctx.DeviceId, ctx.SessionId, ctx.RepositoryPath, record.SongId, record.FilePath,
                    ctx.Options.DryRun, ctx.Options.AutoConfirm,
                    record.Id, record.Reason, ct);

                if (result?.Counts != null)
                {
                    ctx.Result = ctx.Result.AddDelta(result.Counts);
                }
            }
            else if (record.Action == SyncRecordAction.UpdateLocal)
            {
                logger.LogInformation("Updating local song {SongId} at {Path}", record.SongId, record.FilePath);

                var result = await syncActions.ActionUpdateLocalAsync(
                    ctx.DeviceId, ctx.SessionId, ctx.RepositoryPath, record.SongId, record.FilePath,
                    ctx.Options.DryRun, ctx.Options.AutoConfirm,
                    record.Id, record.Reason, ct);

                if (result?.Counts != null)
                {
                    ctx.Result = ctx.Result.AddDelta(result.Counts);
                }
            }
            else if (record.Action == SyncRecordAction.Unlink || record.Action == SyncRecordAction.Delete)
            {
                var result = await syncActions.ActionDeleteAsync(
                    ctx.DeviceId, ctx.SessionId, ctx.RepositoryPath, record.SongId, record.FilePath,
                    ctx.Options.DryRun, ctx.Options.AutoConfirm, record.Id, record.Reason, ct);

                if (result?.Counts != null)
                {
                    ctx.Result = ctx.Result.AddDelta(result.Counts);
                }
            }
            else if (record.Action == SyncRecordAction.Rename)
            {
                var renameData = DeserializeRenameData(record.Data);
                if (renameData != null)
                {
                    var result = await syncActions.ActionRenameAsync(
                        ctx.DeviceId, ctx.SessionId, ctx.RepositoryPath, record.FilePath, renameData.PreviousPath,
                        ctx.Options.DryRun, record.Id, ct);

                    if (result?.Counts != null)
                    {
                        ctx.Result = ctx.Result.AddDelta(result.Counts);
                    }
                }
            }

            progress?.Report(SyncProgress.FromResult(
                ctx.Result, "server", serverTotal, recordsToProcess.IndexOf(record) + 1, record.FilePath));
        }
    }

    public async Task CommitPhaseAsync(
        SyncContext ctx,
        IProgress<SyncProgress>? progress,
        CancellationToken ct = default)
    {
        var directionString = ctx.Options.Direction switch
        {
            SyncDirection.Up => "up",
            SyncDirection.Down => "down",
            _ => "both"
        };

        logger.LogInformation("Committing sync session {SessionId} (direction: {Direction})", ctx.SessionId, directionString);

        var commitResult = await apiClient.CommitSyncAsync(ctx.DeviceId, ctx.SessionId, new CommitSyncRequest
        {
            Direction = directionString
        }, ct);

        logger.LogInformation(
            "Commit result: {CreateRemote} created remote, {UpdateRemote} updated remote, {Skipped} skipped, {CreateLocal} created local, {UpdateLocal} updated local, {Delete} deleted, {Link} linked, {Unlink} unlinked, {Rename} renamed, {Conflict} conflicts, {UpdateTimestamp} timestamps updated, {Error} error",
            commitResult.CreateRemoteCount, commitResult.UpdateRemoteCount, commitResult.SkippedCount,
            commitResult.CreateLocalCount, commitResult.UpdateLocalCount, commitResult.DeleteCount,
            commitResult.LinkCount, commitResult.UnlinkCount, commitResult.RenameCount,
            commitResult.ConflictCount, commitResult.UpdateTimestampCount,
            commitResult.ErrorCount);

        ctx.Result = ctx.Result with
        {
            CreateRemote = commitResult.CreateRemoteCount,
            UpdateRemote = commitResult.UpdateRemoteCount,
            Skipped = commitResult.SkippedCount,
            CreateLocal = commitResult.CreateLocalCount,
            UpdateLocal = commitResult.UpdateLocalCount,
            Delete = commitResult.DeleteCount,
            Link = commitResult.LinkCount,
            Unlink = commitResult.UnlinkCount,
            Rename = commitResult.RenameCount,
            Conflict = commitResult.ConflictCount,
            UpdateTimestamp = commitResult.UpdateTimestampCount,
            Error = commitResult.ErrorCount
        };

        progress?.Report(SyncProgress.FromResult(
            ctx.Result, "commit", 1, 1));
    }

    public async Task CompleteAsync(
        SyncContext ctx,
        int filesCount,
        CancellationToken ct = default)
    {
        var directionString = ctx.Options.Direction switch
        {
            SyncDirection.Up => "up",
            SyncDirection.Down => "down",
            _ => "both"
        };

        var completeResponse = await apiClient.CompleteSyncAsync(ctx.DeviceId, ctx.SessionId, new CompleteSyncRequest
        {
            Direction = directionString
        }, ct);

        logger.LogInformation(
            "Sync complete: {CreateRemote} created remote, {UpdateRemote} updated remote, {Skipped} skipped, {CreateLocal} created local, {UpdateLocal} updated local, {Delete} deleted, {Link} linked, {Unlink} unlinked, {Rename} renamed, {Conflict} conflicts, {UpdateTimestamp} timestamps updated, {Error} error",
            completeResponse.CreateRemoteCount, completeResponse.UpdateRemoteCount, completeResponse.SkippedCount,
            completeResponse.CreateLocalCount, completeResponse.UpdateLocalCount, completeResponse.DeleteCount,
            completeResponse.LinkCount, completeResponse.UnlinkCount, completeResponse.RenameCount,
            completeResponse.ConflictCount, completeResponse.UpdateTimestampCount,
            completeResponse.ErrorCount);

        ctx.Result = ctx.Result with
        {
            CreateRemote = completeResponse.CreateRemoteCount,
            UpdateRemote = completeResponse.UpdateRemoteCount,
            Skipped = completeResponse.SkippedCount,
            CreateLocal = completeResponse.CreateLocalCount,
            UpdateLocal = completeResponse.UpdateLocalCount,
            Delete = completeResponse.DeleteCount,
            Link = completeResponse.LinkCount,
            Unlink = completeResponse.UnlinkCount,
            Rename = completeResponse.RenameCount,
            Conflict = completeResponse.ConflictCount,
            UpdateTimestamp = completeResponse.UpdateTimestampCount,
            Error = completeResponse.ErrorCount
        };
    }

    private static SyncCheckCreateUpdateData? DeserializeCheckCreateUpdateData(System.Text.Json.JsonElement? data)
    {
        if (!data.HasValue || data.Value.ValueKind == System.Text.Json.JsonValueKind.Null)
            return null;
        try
        {
            return System.Text.Json.JsonSerializer.Deserialize<SyncCheckCreateUpdateData>(data.Value, RenameDataOptions);
        }
        catch
        {
            return null;
        }
    }

    internal record SyncCheckCreateUpdateData
    {
        public required DateTime ModifiedAt { get; init; }
        public required DateTime CreatedAt { get; init; }
        public string? Reason { get; init; }
    }

    private static readonly JsonSerializerOptions RenameDataOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    private static RenameData? DeserializeRenameData(JsonElement? data)
    {
        if (!data.HasValue || data.Value.ValueKind == JsonValueKind.Null)
            return null;
        try
        {
            return JsonSerializer.Deserialize<RenameData>(data.Value, RenameDataOptions);
        }
        catch
        {
            return null;
        }
    }
}

internal record RenameData
{
    public required string PreviousPath { get; init; }
    public required string NewPath { get; init; }
}